### ISSUE TITLE
Emerald analyzer stops silently

### ISSUE THREAD

**Issue by buberdds** [oasisprotocol/nexus#267]
https://index.oasislabs.com/v1/emerald/transactions?limit=10&offset=0 the latest indexed block is 3934708

**Comment by mitjat** [comment]
Thanks for reporting.

I had to dig pretty far into log history to find any entries pertaining to emerald. The most recent entry reads `{..., "msg":"processing round", "round":3934709}`. Followed by no errors, just radio silence. I also checked for errors, core dumps etc not associated with the emerald_damask analyzer, but there's nothing.

Very weird. :confused: The analyzer runs an infinite main loop that should print out an error or else move on to the next block ...

Restarting the analyzer fixed it. Slightly unsettling, but I don't think we should dig further for now, given this is the first time it happened. Also, I have no idea how to try and repro, other than running a bunch of stacks and waiting. The next version of analyzer (merged in git, but not yet deployed) has a little more logging, hopefully that will help, should the problem recur.

**Comment by mitjat** [comment]
Reopening because we just saw two more cases of this within a day (one in staging, one in prod; both at different heights; both running v0.0.5 release).

Early debugging. Produced stacktrace dump with
```
kubectl -n mainnet exec -i -t deployments/mainnet-oasis-indexer -c analyzer -- bash -c 'kill -QUIT 1'
```
(This dumps stack traces AND kills the program; k8s restarts it immediately. Btoh staging and prod recovered nicely.)

Looked at stderr in GCP or with
```
kubectl -n mainnet logs deployments/mainnet-oasis-indexer -c analyzer --previous --since 1h| less
```

Here are all the goroutines that contain "emerald" in the stack trace ...
<details>
<summary>Goroutines</summary>

```
goroutine 259858 [chan receive, 1153 minutes]:
google.golang.org/grpc/internal/transport.(*http2Client).closeStream(0x16c6c60?, 0xc00019a088?, {0x16c6c60?, 0xc00019a088?}, 0xb0?, 0xc0?, 0xb857a8?, 0x16c6c60?, 0x88?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/http2_client.go:822 +0x22d
google.golang.org/grpc/internal/transport.(*http2Client).CloseStream(0x3e?, 0xc000c75548?, {0x16c6c60, 0xc00019a088})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/http2_client.go:814 +0x6b
google.golang.org/grpc.(*csAttempt).finish(0xc00100d380, {0x16c6c60, 0xc00019a088})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:1060 +0x126
google.golang.org/grpc.(*clientStream).retryLocked(0xc000a24fc0, 0x0?, {0x16c6c60?, 0xc00019a088?})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:659 +0x65
google.golang.org/grpc.(*clientStream).withRetry(0xc000a24fc0, 0xc000c75810, 0xc000c757e0)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:729 +0x293
google.golang.org/grpc.(*clientStream).RecvMsg(0xc000a24fc0, {0x10a5620?, 0xc00052b470?})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:850 +0x11f
google.golang.org/grpc.invoke({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0xc00043e5a0?}, {0x10b31a0, 0xc000979ef0}, {0x10a5620, 0xc00052b470}, 0x40da0d?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:73 +0xd7
github.com/oasisprotocol/oasis-core/go/common/grpc.clientUnaryErrorMapper({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x11eb2a0?}, {0x10b31a0?, 0xc000979ef0?}, {0x10a5620?, 0xc00052b470?}, 0xc0002568a0?, 0x157ab48, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/errors.go:126 +0x7e
google.golang.org/grpc.getChainUnaryInvoker.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc000979ef0}, {0x10a5620, 0xc00052b470}, 0x203000?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:361 +0x154
github.com/oasisprotocol/oasis-core/go/common/grpc.(*grpcLogAdapter).unaryClientLogger(0xc000374810, {0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0?, 0xc000979ef0}, {0x10a5620?, 0xc00052b470?}, 0xc000498380, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/grpc.go:231 +0x337
google.golang.org/grpc.chainUnaryClientInterceptors.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc000979ef0}, {0x10a5620, 0xc00052b470}, 0x1002af511e6f0?, 0x157ab48, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:349 +0x128
google.golang.org/grpc.(*ClientConn).Invoke(0x7f2b1f9eaa68?, {0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x0?}, {0x10b31a0?, 0xc000979ef0?}, {0x10a5620?, 0xc00052b470?}, {0x0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:35 +0x223
github.com/oasisprotocol/oasis-core/go/runtime/client/api.(*runtimeClient).GetEvents(0xc000308058, {0x16d18e0, 0xc001214ec0}, 0x41031e?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/runtime/client/api/grpc.go:500 +0xa7
github.com/oasisprotocol/oasis-sdk/client-sdk/go/client.(*runtimeClient).GetEventsRaw(0xc000472190, {0x16d18e0, 0xc001214ec0}, 0x1019d4)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/client/client.go:354 +0xa9
github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules/core.(*v1).GetEvents(0x4bdf4d?, {0x16d18e0?, 0xc001214ec0?}, 0xc0009d1e48?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/modules/core/core.go:102 +0x43
github.com/oasisprotocol/oasis-indexer/storage/oasis.(*RuntimeClient).CoreData(0x3755?, {0x16d18e0?, 0xc001214ec0?}, 0x1019d4)
        /code/go/storage/oasis/runtime.go:49 +0x3e
github.com/oasisprotocol/oasis-indexer/analyzer/modules.(*CoreHandler).PrepareData(0xc0004f6120, {0x16d18e0?, 0xc001214ec0?}, 0x3755?, 0x4000?)
        /code/go/analyzer/modules/core.go:32 +0x42
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).processRound.func2.1()
        /code/go/analyzer/emerald/emerald.go:234 +0x2d
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5


goroutine 73 [semacquire, 1153 minutes]:
sync.runtime_Semacquire(0xc000442b60?)
        /usr/local/go/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc001214f01?)
        /usr/local/go/src/sync/waitgroup.go:136 +0x52
golang.org/x/sync/errgroup.(*Group).Wait(0xc001214f00)
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:53 +0x27
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).processRound(0xc0001b2200, {0x16d1918?, 0xc00004c030}, 0x1019d4)
        /code/go/analyzer/emerald/emerald.go:249 +0x2d9
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).Start(0xc0001b2200)
        /code/go/analyzer/emerald/emerald.go:151 +0x377
github.com/oasisprotocol/oasis-indexer/cmd/analyzer.(*Service).Start.func1({0x16caf98?, 0xc0001b2200?})
        /code/go/cmd/analyzer/analyzer.go:194 +0x66
created by github.com/oasisprotocol/oasis-indexer/cmd/analyzer.(*Service).Start


goroutine 259859 [semacquire, 1153 minutes]:
sync.runtime_SemacquireMutex(0xc0011cdba8?, 0x0?, 0x203000?)
        /usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc000000340)
        /usr/local/go/src/sync/mutex.go:162 +0x165
sync.(*Mutex).Lock(...)
        /usr/local/go/src/sync/mutex.go:81
google.golang.org/grpc/internal/transport.(*http2Client).NewStream.func3({0x113e220?, 0xc000772a50?})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/http2_client.go:723 +0x11d
google.golang.org/grpc/internal/transport.(*http2Client).NewStream.func5({0x113e220, 0xc000772a50})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/http2_client.go:755 +0x4c
google.golang.org/grpc/internal/transport.(*controlBuffer).executeAndPut(0xc000bbad20, 0x16d1988?, {0x16c6d60, 0xc000772a50})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/controlbuf.go:337 +0xc3
google.golang.org/grpc/internal/transport.(*http2Client).NewStream(0xc0000001e0, {0x16d1988, 0xc000b9f2f0}, 0xc001114480)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/http2_client.go:754 +0x432
google.golang.org/grpc.(*csAttempt).newStream(0xc000c79d40)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:441 +0x4a
google.golang.org/grpc.newClientStreamWithParams.func2(0xc000c79d40)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:312 +0x3a
google.golang.org/grpc.(*clientStream).withRetry(0xc000ae06c0, 0xc000c9a820, 0xc000c21628)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:715 +0x144
google.golang.org/grpc.newClientStreamWithParams({0x16d18e0, 0xc001214ec0}, 0x1f3e880, 0xc000498380, {0xc0001cd860, 0x23}, {0x0, 0x0, 0x0, 0x0, ...}, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:321 +0x98c
google.golang.org/grpc.newClientStream.func2({0x16d18e0?, 0xc001214ec0?}, 0xc001214ec0?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:192 +0x9f
google.golang.org/grpc.newClientStream({0x16d18e0, 0xc001214ec0}, 0x1f3e880, 0xc000498380, {0xc0001cd860, 0x23}, {0xc0003748d0, 0x3, 0x3})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:220 +0x4c2
google.golang.org/grpc.invoke({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0xc00043e5a0?}, {0x10b31a0, 0xc000ac09f0}, {0x10a5620, 0xc000467b48}, 0x40da0d?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:66 +0x7d
github.com/oasisprotocol/oasis-core/go/common/grpc.clientUnaryErrorMapper({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x11eb2a0?}, {0x10b31a0?, 0xc000ac09f0?}, {0x10a5620?, 0xc000467b48?}, 0xc0002568a0?, 0x157ab48, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/errors.go:126 +0x7e
google.golang.org/grpc.getChainUnaryInvoker.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc000ac09f0}, {0x10a5620, 0xc000467b48}, 0x203000?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:361 +0x154
github.com/oasisprotocol/oasis-core/go/common/grpc.(*grpcLogAdapter).unaryClientLogger(0xc000374810, {0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0?, 0xc000ac09f0}, {0x10a5620?, 0xc000467b48?}, 0xc000498380, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/grpc.go:231 +0x337
google.golang.org/grpc.chainUnaryClientInterceptors.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc000ac09f0}, {0x10a5620, 0xc000467b48}, 0x16d18e0?, 0x157ab48, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:349 +0x128
google.golang.org/grpc.(*ClientConn).Invoke(0x7f2b1f9ea5b8?, {0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x0?}, {0x10b31a0?, 0xc000ac09f0?}, {0x10a5620?, 0xc000467b48?}, {0x0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:35 +0x223
github.com/oasisprotocol/oasis-core/go/runtime/client/api.(*runtimeClient).GetEvents(0xc000308058, {0x16d18e0, 0xc001214ec0}, 0x28?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/runtime/client/api/grpc.go:500 +0xa7
github.com/oasisprotocol/oasis-sdk/client-sdk/go/client.(*runtimeClient).GetEventsRaw(0xc000472190, {0x16d18e0, 0xc001214ec0}, 0x1019d4)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/client/client.go:354 +0xa9
github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules/accounts.(*v1).GetEvents(0x4461d1?, {0x16d18e0?, 0xc001214ec0?}, 0xc0006da820?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/modules/accounts/accounts.go:115 +0x43
github.com/oasisprotocol/oasis-indexer/storage/oasis.(*RuntimeClient).AccountsData(0x4dbf44?, {0x16d18e0?, 0xc001214ec0?}, 0x1019d4)
        /code/go/storage/oasis/runtime.go:72 +0x4c
github.com/oasisprotocol/oasis-indexer/analyzer/modules.(*AccountsHandler).PrepareData(0xc0004f6240, {0x16d18e0?, 0xc001214ec0?}, 0x8199?, 0xa000?)
        /code/go/analyzer/modules/accounts.go:32 +0x53
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).processRound.func2.1()
        /code/go/analyzer/emerald/emerald.go:234 +0x2d
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5


goroutine 259860 [select, 1153 minutes]:
google.golang.org/grpc/internal/transport.(*Stream).waitOnHeader(0xc0005e8000)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/transport.go:324 +0x7c
google.golang.org/grpc/internal/transport.(*Stream).RecvCompress(...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/transport.go:339
google.golang.org/grpc.(*csAttempt).recvMsg(0xc000bf1a40, {0x10a5620?, 0xc000f89080}, 0x7f2af5a65068?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:986 +0xbe
google.golang.org/grpc.(*clientStream).RecvMsg.func1(0x0?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:851 +0x25
google.golang.org/grpc.(*clientStream).withRetry(0xc0005c7d40, 0xc000bbd7d0, 0xc000bbd7a0)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:715 +0x144
google.golang.org/grpc.(*clientStream).RecvMsg(0xc0005c7d40, {0x10a5620?, 0xc000f89080?})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:850 +0x11f
google.golang.org/grpc.invoke({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0xc00043e5a0?}, {0x10b31a0, 0xc0005b7ce0}, {0x10a5620, 0xc000f89080}, 0x40da0d?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:73 +0xd7
github.com/oasisprotocol/oasis-core/go/common/grpc.clientUnaryErrorMapper({0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x11eb2a0?}, {0x10b31a0?, 0xc0005b7ce0?}, {0x10a5620?, 0xc000f89080?}, 0xc0002568a0?, 0x157ab48, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/errors.go:126 +0x7e
google.golang.org/grpc.getChainUnaryInvoker.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc0005b7ce0}, {0x10a5620, 0xc000f89080}, 0x203000?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:361 +0x154
github.com/oasisprotocol/oasis-core/go/common/grpc.(*grpcLogAdapter).unaryClientLogger(0xc000374810, {0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0?, 0xc0005b7ce0}, {0x10a5620?, 0xc000f89080?}, 0xc000498380, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/grpc.go:231 +0x337
google.golang.org/grpc.chainUnaryClientInterceptors.func1({0x16d18e0, 0xc001214ec0}, {0xc0001cd860, 0x23}, {0x10b31a0, 0xc0005b7ce0}, {0x10a5620, 0xc000f89080}, 0x7f?, 0x157ab48, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:349 +0x128
google.golang.org/grpc.(*ClientConn).Invoke(0x7f2b1f9ea108?, {0x16d18e0?, 0xc001214ec0?}, {0xc0001cd860?, 0x0?}, {0x10b31a0?, 0xc0005b7ce0?}, {0x10a5620?, 0xc000f89080?}, {0x0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:35 +0x223
github.com/oasisprotocol/oasis-core/go/runtime/client/api.(*runtimeClient).GetEvents(0xc000308058, {0x16d18e0, 0xc001214ec0}, 0x28?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/runtime/client/api/grpc.go:500 +0xa7
github.com/oasisprotocol/oasis-sdk/client-sdk/go/client.(*runtimeClient).GetEventsRaw(0xc000472190, {0x16d18e0, 0xc001214ec0}, 0x1019d4)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/client/client.go:354 +0xa9
github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules/consensusaccounts.(*v1).GetEvents(0xc0009cce00?, {0x16d18e0?, 0xc001214ec0?}, 0x203000?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/modules/consensusaccounts/consensus_accounts.go:100 +0x43
github.com/oasisprotocol/oasis-indexer/storage/oasis.(*RuntimeClient).ConsensusAccountsData(0x8?, {0x16d18e0?, 0xc001214ec0?}, 0x1019d4)
        /code/go/storage/oasis/runtime.go:103 +0x4c
github.com/oasisprotocol/oasis-indexer/analyzer/modules.(*ConsensusAccountsHandler).PrepareData(0xc0004f6260, {0x16d18e0?, 0xc001214ec0?}, 0x74?, 0xc0009ccf50?)
        /code/go/analyzer/modules/consensusaccounts.go:32 +0x53
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).processRound.func2.1()
        /code/go/analyzer/emerald/emerald.go:234 +0x2d
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5



goroutine 259793 [select, 1153 minutes]:
google.golang.org/grpc/internal/transport.(*Stream).waitOnHeader(0xc000b10240)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/transport.go:324 +0x7c
google.golang.org/grpc/internal/transport.(*Stream).RecvCompress(...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/internal/transport/transport.go:339
google.golang.org/grpc.(*csAttempt).recvMsg(0xc00011fb00, {0x10b2ea0?, 0xc000b06000}, 0x7f2af5a65068?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:986 +0xbe
google.golang.org/grpc.(*clientStream).RecvMsg.func1(0x0?)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:851 +0x25
google.golang.org/grpc.(*clientStream).withRetry(0xc000b10000, 0xc00093d998, 0xc00093d968)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:715 +0x144
google.golang.org/grpc.(*clientStream).RecvMsg(0xc000b10000, {0x10b2ea0?, 0xc000b06000?})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/stream.go:850 +0x11f
google.golang.org/grpc.invoke({0x16d1918?, 0xc00004c030?}, {0xc0001cd740?, 0xc00043e300?}, {0x10b3160, 0xc000d37650}, {0x10b2ea0, 0xc000b06000}, 0x40da0d?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:73 +0xd7
github.com/oasisprotocol/oasis-core/go/common/grpc.clientUnaryErrorMapper({0x16d1918?, 0xc00004c030?}, {0xc0001cd740?, 0x11eb2a0?}, {0x10b3160?, 0xc000d37650?}, {0x10b2ea0?, 0xc000b06000?}, 0xc0002568a0?, 0x157ab48, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/errors.go:126 +0x7e
google.golang.org/grpc.getChainUnaryInvoker.func1({0x16d1918, 0xc00004c030}, {0xc0001cd740, 0x22}, {0x10b3160, 0xc000d37650}, {0x10b2ea0, 0xc000b06000}, 0x203000?, {0xc0003748d0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:361 +0x154
github.com/oasisprotocol/oasis-core/go/common/grpc.(*grpcLogAdapter).unaryClientLogger(0xc000374810, {0x16d1918, 0xc00004c030}, {0xc0001cd740, 0x22}, {0x10b3160?, 0xc000d37650}, {0x10b2ea0?, 0xc000b06000?}, 0xc000498380, ...)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/common/grpc/grpc.go:231 +0x337
google.golang.org/grpc.chainUnaryClientInterceptors.func1({0x16d1918, 0xc00004c030}, {0xc0001cd740, 0x22}, {0x10b3160, 0xc000d37650}, {0x10b2ea0, 0xc000b06000}, 0x0?, 0x157ab48, ...)
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/clientconn.go:349 +0x128
google.golang.org/grpc.(*ClientConn).Invoke(0x7f2b1f9eaf18?, {0x16d1918?, 0xc00004c030?}, {0xc0001cd740?, 0x0?}, {0x10b3160?, 0xc000d37650?}, {0x10b2ea0?, 0xc000b06000?}, {0x0, ...})
        /go/pkg/mod/google.golang.org/grpc@v1.49.0/call.go:35 +0x223
github.com/oasisprotocol/oasis-core/go/runtime/client/api.(*runtimeClient).GetBlock(0xc000308058, {0x16d1918, 0xc00004c030}, 0xc0009cd700?)
        /go/pkg/mod/github.com/oasisprotocol/oasis-core/go@v0.2202.3/runtime/client/api/grpc.go:468 +0xa5
github.com/oasisprotocol/oasis-sdk/client-sdk/go/client.(*runtimeClient).GetBlock(0xc000472190, {0x16d1918, 0xc00004c030}, 0x1019d4)
        /go/pkg/mod/github.com/oasisprotocol/oasis-sdk/client-sdk/go@v0.2.1-0.20220823181727-bd5025a9857b/client/client.go:292 +0x82
github.com/oasisprotocol/oasis-indexer/storage/oasis.(*RuntimeClient).BlockData(0xc0001b2100, {0x16d1918, 0xc00004c030}, 0x1019d4)
        /code/go/storage/oasis/runtime.go:30 +0x3c
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).prepareBlockData(0xc0001b2200, {0x16d1918, 0xc00004c030}, 0x54b8e5?, 0xc0001b2a80?)
        /code/go/analyzer/emerald/emerald.go:270 +0x3d
github.com/oasisprotocol/oasis-indexer/analyzer/emerald.(*Main).processRound.func1()
        /code/go/analyzer/emerald/emerald.go:227 +0x2d
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5
```

</details>

**Comment by mitjat** [comment]
All the other emerald-related goroutines seem to just be in a hung gRPC connection talking to the node. Is it possible that oasis-node can deadlock in case of too many simultaneous connections? Why is this behavior starting only now; the "problematic" (if they are that) node queries are just `GetEvents` and `GetBlock()`, which we've been using forever. We're not doing fancy EVM calls here to glean ERC-20 token info.

**Comment by pro-wh** [comment]
I'm going to try opening a PR with timeouts on the GRPC calls. which is a little more conservative than having it on the whole `process`-* method. the database access could still hang our loop, but we don't have intel on how that is likely to fail

**Comment by mitjat** [comment]
@kostko wondered if it might be oasis-node that was causing issues.

I believe the oasis-node running now in prod is still the same process that was serving the hung-up indexer.
I dumped its goroutines (`kubectl -n mainnet exec -i -t pods/mainnet-grpc-v2202-oasis-client-0 -- bash -c 'kill -QUIT $(pidof oasis-node)'`), and only two have `grpc` in their stack trace:

<details>
<summary>Stack traces of grpc gorotuines</summary>

```
goroutine 8535 [IO wait]:
internal/poll.runtime_pollWait(0x7f589a2b7800, 0x72)
	runtime/netpoll.go:302 +0x89
internal/poll.(*pollDesc).wait(0xc011588e80?, 0x18?, 0x0)
	internal/poll/fd_poll_runtime.go:83 +0x32
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:88
internal/poll.(*FD).Accept(0xc011588e80)
	internal/poll/fd_unix.go:614 +0x22c
net.(*netFD).accept(0xc011588e80)
	net/fd_unix.go:172 +0x35
net.(*UnixListener).accept(0x4492e0?)
	net/unixsock_posix.go:166 +0x1c
net.(*UnixListener).Accept(0xc01349cc60)
	net/unixsock.go:260 +0x3d
google.golang.org/grpc.(*Server).Serve(0xc0000010e0, {0x217c560?, 0xc01349cc60})
	google.golang.org/grpc@v1.49.0/server.go:796 +0x477
github.com/oasisprotocol/oasis-core/go/common/grpc.(*Server).Start.func1()
	github.com/oasisprotocol/oasis-core/go/common/grpc/grpc.go:487 +0x72
created by github.com/oasisprotocol/oasis-core/go/common/grpc.(*Server).Start
	github.com/oasisprotocol/oasis-core/go/common/grpc/grpc.go:484 +0x4f5

goroutine 8536 [semacquire, 48528 minutes]:
sync.runtime_Semacquire(0x401f0f177407e1?)
	runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0x444e4554584545?)
	sync/waitgroup.go:136 +0x52
github.com/oasisprotocol/oasis-core/go/common/grpc.(*Server).Start.func2()
	github.com/oasisprotocol/oasis-core/go/common/grpc/grpc.go:496 +0x26
created by github.com/oasisprotocol/oasis-core/go/common/grpc.(*Server).Start
	github.com/oasisprotocol/oasis-core/go/common/grpc/grpc.go:495 +0x1fb
```

</details>


That doesn't look off to me. Looks like one goroutine waiting for grpc connections, and one oasis-core wrapper waiting for the grpc library. 

For completeness, here's [the full goroutine dump](https://github.com/oasisprotocol/oasis-indexer/files/10382372/goroutines.txt)

TLDR: Nothing looks off on the oasis-core side from my preliminary investigation.

**Comment by ptrus** [comment]
This was resolved.
